tools/mkcomposefs.c

```diff
@@ -240,20 +240,17 @@ static int copy_file_with_dirs_if_needed(const char *src, const char *dst_base,
 		if (res < 0) {
 			errsv = errno;
 			unlink(tmppath);
-			close(sfd);
-			close(dfd);
 			errno = errsv;
 			return res;
 		}
 	}
-	close(sfd);
+	close(sfd); sfd = -1;
```

Comment:
Well I just realized something...because all of the code here so far relies on propagating errno, we need something like the whole [PROTECT_ERRNO](https://github.com/systemd/systemd/blob/6e89663d23230bef4ddcbb0a80bfdf6daa714887/src/basic/errno-util.h#L31) infrastructure and to use it in our `cleanup_close`.